### PR TITLE
feat(router): angular1 router

### PR DESCRIPTION
**Pull request by btford**
Supersedes #3138

I did a bit of cleanup on @shahata's PR. I'm going to let CI for this run, and push it to presubmit if it passes.

@IgorMinar is going to review this, but we should get this in before it gets bitrotten again.

### CODE REVIEW

File: modules/angular1_router/lib/facades.es5 [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+  create: function () {
+    return {};
+  },
+
+  set: function (m, k, v) {
+    return m[k] = v;
+  },
+
+  get: function (m, k) {
+    return m[k];
```
Comment (gkalpak) [inline modules/angular1_router/lib/facades.es5]: Continuing the discussion from https://github.com/angular/angular/pull/3138#discussion_r37337385, `get()`, `contains()` and `isEmpty()` should use `hasOwnProperty()` (unless the objects passed to it have no prototype).

Even `facade/collection.ts`, uses `hasOwnProperty()`.

E.g. `isEmpty()` will never return truw, if there are properties on the prototype.

Comment (btford) [inline modules/angular1_router/lib/facades.es5]: Ah – this is legit. :+1:

File: modules/angular1_router/test/ng_outlet_spec.js [hunk 2]
```diff
@@ -0,0 +1,889 @@
...
+
+    $router.navigate('/');
+    $rootScope.$digest();
+
+    expect(elt.text()).toBe('one');
+  });
+
+
+  // See https://github.com/angular/router/issues/105
+  xit('should warn when instantiating a component with no controller', function () {
```
Comment (IgorMinar) [inline modules/angular1_router/test/ng_outlet_spec.js]: xit-ed test - is this still correct? the issue listed above is from the wrong repo

File: modules/angular1_router/test/ng_outlet_spec.js [hunk 3]
```diff
@@ -0,0 +1,889 @@
...
+    expect(elt.text()).toBe('hello brian');
+
+    $router.navigate('/user/igor');
+    $rootScope.$digest();
+    expect(elt.text()).toBe('hello igor');
+  });
+
+
+  //TODO: fix this
+  xit('should not reactivate a parent when navigating between child components with different parameters', function () {
```
Comment (IgorMinar) [inline modules/angular1_router/test/ng_outlet_spec.js]: xit

File: modules/angular1_router/test/ng_outlet_spec.js [hunk 4]
```diff
@@ -0,0 +1,889 @@
...
+    spy.calls.reset();
+
+    $router.navigate('/parent/user/igor');
+    $rootScope.$digest();
+    expect(spy).not.toHaveBeenCalled();
+    expect(elt.text()).toBe('parent { hello igor }');
+  });
+
+
+  // it('should work with multiple named outlets', function () {
```
Comment (IgorMinar) [inline modules/angular1_router/test/ng_outlet_spec.js]: commented out

File: modules/angular1_router/test/ng_outlet_spec.js [hunk 5]
```diff
@@ -0,0 +1,889 @@
...
+    $rootScope.$digest();
+    expect(elt.text()).toBe('link | one');
+    elt.find('a')[0].click();
+
+    $rootScope.$digest();
+    expect(elt.text()).toBe('link | two');
+  });
+
+
+  xit('should not navigate on non-left mouse click when a link url matches a route', inject(function ($router) {
```
Comment (IgorMinar) [inline modules/angular1_router/test/ng_outlet_spec.js]: xit

File: modules/angular1_router/test/ng_outlet_spec.js (hunk 4 above)
Comment (btford) [inline modules/angular1_router/test/ng_outlet_spec.js]: :+1: This should be removed and re-implemented.

File: modules/angular1_router/test/ng_outlet_spec.js (hunk 2 above)
Comment (btford) [inline modules/angular1_router/test/ng_outlet_spec.js]: yes– this behavior needs to be re-thought regardless.

**Comment by btford** [comment]
I've removed all but one `xit`.

I'm hoping we can get this in today and then I'll spend the rest of the week improving tests and refactoring.

**Comment by btford** [comment]
@IgorMinar gave me the LGTM to merge this with the understanding that I'll follow up with some refactor and cleanup work.

Pushing to presubmit :cat2:

**Comment by vicb** [comment]
@btford I'm re-opening this PR as it had to be reverted (Travis was red)

**Comment by btford** [comment]
This PR clearly was not the cause of Travis going red. After reverting this commit, we still see the same failure.

@vicb can you re-merge this please?

**Comment by btford** [comment]
Reverted @vicb's revert. This is back in master.

@vicb – next time please make sure CI passes before pushing to changes to master.

**Comment by poshest** [comment]
What's the status of $router in 1.x branch?  Will we have it for 1.5?  I'm just getting to the point where I need to put a router into my app, and would like some guidance about whether to wait or just go with ngRoute or ui-router for now.

**Pull request closed without merging**